server/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -82,6 +98,14 @@ public AllocateUnassignedDecision makeAllocationDecision(final ShardRouting unas
             return AllocateUnassignedDecision.NOT_TAKEN;
         }
 
+        final IndexMetaData indexMetaData = allocation.metaData().getIndexSafe(unassignedShard.index());
+        final String existingShardsAllocator = ExistingShardsAllocator.EXISTING_SHARDS_ALLOCATOR_SETTING.get(indexMetaData.getSettings());
+        if (unassignedShard.recoverySource().getType() == RecoverySource.Type.EXISTING_STORE
+            && existingShardsAllocator.equals(GatewayAllocator.ALLOCATOR_NAME) == false) {
```

Comment:
I don't really understand this suggestion, sorry. Are you saying that we should allocate the shard as normal if the plugin is missing?

Reply:
I wonder if we should have a separate facility that ensures that certain plugins are installed in the presence of certain types of indices. Otherwise we have to build this again and again for any plugin-specific setting that we introduce.

Take for example "follower" indices (CCR). Those only make sense in the context of x-pack.

Reply:
I now see the leniency in `MetaDataIndexUpgradeService.archiveBrokenIndexSettings`, I thought we would be stricter on that. It seems dangerous that we just archive all the settings and then act like nothing happened. I think not allocating any such indices until resolved would be better, but that is likely a more intrusive change.

Maybe there is a different way we can deem indices with searchable snapshot settings without the underlying plugin more invalid than other superfluous settings?

Reply:
I have reverted these changes to the `PrimaryShardAllocator` and instead now the `AllocationService` chooses the correct allocator based on an index setting. I think a setting is a good approach; we could perhaps deduce the fact that a different allocator is needed for searchable snapshots via some other mechanism, but I'm not sure what that mechanism might be nor what advantage it offers.

I am generally in favour of validating that the nodes in the cluster "make sense" in terms of their plugins but that's a larger topic for another day IMO.

Reply:
@henningandersen See #26998.

Reply:
My main issue with the approach is redundancy. I can see how the redundancy here solves the issue of a missing plugin and agree that solving that is not now.